Populate endpoints for headless service with no ports

xiangpengzhao opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:
- populate endpoints with headless service (thanks @fraenkel for the original PR!)
- allow ports with headless service
- nits

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #32796 https://github.com/kubernetes/kubernetes/issues/32796#issuecomment-270462724

**Special notes for your reviewer**:
/cc @thockin @fraenkel 
**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
NONE
```

xiangpengzhao committed:
Populate endpoints and allow ports with headless service

thockin commented:
@k8s-bot ok to test

thockin commented on pkg/kubectl/cmd/create_service.go:
```diff
@@ -48,25 +48,25 @@ func NewCmdCreateService(f cmdutil.Factory, cmdOut, errOut io.Writer) *cobra.Com
 }
 
-// NewCmdCreateServiceClusterIP is a command to create a clusterIP service
+// NewCmdCreateServiceClusterIP is a command to create a ClusterIP service
```
In the future, please move such cleanups to a different PR or at least a different commit in the same PR

xiangpengzhao replied:
Thanks Tim for the advice!

thockin commented on pkg/kubectl/service_basic.go:
```diff
@@ -189,9 +189,6 @@ func (s ServiceCommonGeneratorV1) validate() error {
 	if s.ClusterIP == api.ClusterIPNone && s.Type != api.ServiceTypeClusterIP {
 		return fmt.Errorf("ClusterIP=None can only be used with ClusterIP service type")
 	}
-	if s.ClusterIP == api.ClusterIPNone && len(s.TCP) > 0 {
```
we should still be asserting that any ports specified are identity mapped.

xiangpengzhao replied:
Tim, what do you mean by `identity mapped`? Is it that port and targetPort should be equal for headless service, right?

xiangpengzhao replied:
API validation doesn't validate that the port and targetPort should be equal for headless service. The code is commented currently.

https://github.com/kubernetes/kubernetes/blob/master/pkg/api/validation/validation.go#L2866-L2873
```
	// in the v1 API, targetPorts on headless services were tolerated.
	// once we have version-specific validation, we can reject this on newer API versions, but until then, we have to tolerate it for compatibility.
	//
	// if isHeadlessService {
	// 	if sp.TargetPort.Type == intstr.String || (sp.TargetPort.Type == intstr.Int && sp.Port != sp.TargetPort.IntValue()) {
	// 		allErrs = append(allErrs, field.Invalid(fldPath.Child("targetPort"), sp.TargetPort, "must be equal to the value of 'port' when clusterIP = None"))
	// 	}
	// }
```

thockin replied:
ugh, can you just leave a comment, then?

thockin commented on pkg/kubectl/service_basic_test.go:
```diff
@@ -58,13 +58,6 @@ func TestServiceBasicGenerate(t *testing.T) {
 			expectErr:   true,
 		},
 		{
-			name:        "clusterip-none and port mapping",
```
This is still wrong - headless ports should be identity mapped ONLY

xiangpengzhao replied:
Waiting for `identity mapped` and validation code to be confirmed.

thockin replied:
"identity mapped" means X -> X, Y -> Y.  BUt as above, this would be a break, so just comment why we don't enforce it.

thockin reviewed (commented):
This is a great start.  We need to make sure that other code that reads this struct doesn't barf on zero values.

DNS could use a specific unit test.

thockin commented:
I am going to be away from keyboard for a while.  Assigning to @bowei to wrap up.

See specifically https://github.com/kubernetes/kubernetes/issues/32796#issuecomment-247797321 for a list of places to audit.

dims commented:
/unassign

xiangpengzhao commented:
/retest

xiangpengzhao commented:
updated. PTAL. @thockin @bowei 

In this update, inline comments are addressed.  Unrelated fixes are also removed to avoid noise in the PR. 

I also audit the list of places which might be affected.
- [x] kubectl
* modify validation and test of flag `--tcp` (file service_basic.go/service_basic_test.go in this PR)
* add test case for ServiceGeneratorV2 (file service_test.go)
* related func [generate](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/service.go#L131-L140): already respected headless service.
* related func [RunExpose](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/expose.go#L188-L191): already respected headless service.

- [x] API validation
* modify validation and test of Ports of EndpointSubsets (file pkg/api/validation/validation.go and validation_test.go). also add a helper in `staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go`

- [x] API defaulting
* related func [SetDefaults_Endpoints](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/v1/defaults.go#L257): don't need to modify.

- [x] endpoints controller
* respect len(service.Spec.Ports)==0 (file endpoints_controller.go/endpoints_controller_test.go)

- [x] subset repacking 
* this is okay after endpoints controller respects len(service.Spec.Ports)==0.

- [ ] kube-proxy
* modify test case in proxier_test.go
* I might miss something here...

- [ ] DNS server (and "DNS could use a specific unit test"): 
* don't know where to check... need help here

thockin commented on pkg/api/validation/validation.go:
```diff
@@ -3972,9 +3972,10 @@ func validateEndpointSubsets(subsets []api.EndpointSubset, oldSubsets []api.Endp
-			allErrs = append(allErrs, field.Required(idxPath.Child("ports"), ""))
-		}
+		// for headless service, we allow its endpoints not to have ports.
+		//if len(ss.Ports) == 0 {
```
no need to leave the code commented, just the comment is sufficient.  Maybe:

xiangpengzhao replied:
I don't know if it's clear enough when there is only a single comment without context...

thockin commented on staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go:
```diff
@@ -190,6 +190,14 @@ func IsValidPortNum(port int) []string {
 }
 
+// IsValidPortNumWithZeroInclusive tests that the argument is a valid port number, with zero inclusive.
+func IsValidPortNumWithZeroInclusive(port int) []string {
```
This is not really generically useful, can we instead just check "or equals 0" at the callsite?

xiangpengzhao replied:
I thought there might be some other ports which allow 0 value. I remembered that I saw a use case when I reviewed PRs. But I'm okay to remove it here.

thockin commented on pkg/api/validation/validation.go:
```diff
@@ -4044,7 +4045,7 @@ func validateEndpointPort(port *api.EndpointPort, requireName bool, fldPath *fie
 		allErrs = append(allErrs, ValidateDNS1123Label(port.Name, fldPath.Child("name"))...)
 	}
-	for _, msg := range validation.IsValidPortNum(int(port.Port)) {
+	for _, msg := range validation.IsValidPortNumWithZeroInclusive(int(port.Port)) {
```
prefer to just check for zero here, since nobody else will use that function.

thockin replied:
wait, why do we need to allow zero?  If they specified an EPP block, they should be forced to specify a valid port, right?

xiangpengzhao replied:
>For headless services that do not define selectors, the endpoints controller does not create Endpoints records. However, the DNS system looks for and configures either:
CNAME records for ExternalName-type services
A records for any Endpoints that share a name with the service, for all other types

As per https://kubernetes.io/docs/concepts/services-networking/service/#without-selectors, when users configure a headless service without ports, they should create endpoints themselves. If they don't specify ports in endpoint manifest, it's valid (as we remove `if len(ss.Ports) == 0` in L3976). But if they specify ports (with zero value), should we allow this? I might misunderstand here...

xiangpengzhao commented:
@thockin I reply inline comments. PTAL. Thanks!

xiangpengzhao commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

bowei commented:
/assign @freehan

freehan commented on pkg/controller/endpoint/endpoints_controller.go:
```diff
@@ -338,61 +338,68 @@ func (e *EndpointController) syncService(key string) error {
 
 	readyEps := 0
 	notReadyEps := 0
+	addEndpointSubset := func(pod *v1.Pod, epa v1.EndpointAddress, epp v1.EndpointPort) {
```
I really do not like to have a function variable and let it manipulate local vars in line. Can we separate this into 2 functions? Maybe add another entry function on top of them?

I am fine with any refactoring that can make the logic obvious and straightforward. That will also be good for unit testing.

freehan commented on pkg/controller/endpoint/endpoints_controller.go:
```diff
@@ -338,61 +338,68 @@ func (e *EndpointController) syncService(key string) error {
-					Ports:             []v1.EndpointPort{epp},
-				})
-				notReadyEps++
+		if len(service.Spec.Ports) == 0 {
```
add a comment to say this is for headless service

xiangpengzhao commented:
inline comments addressed. PTAL. @freehan

xiangpengzhao commented:
/retest

freehan commented on pkg/controller/endpoint/endpoints_controller.go:
```diff
@@ -336,63 +335,59 @@ func (e *EndpointController) syncService(key string) error {
+				}
+
+				epp := v1.EndpointPort{Name: portName, Port: int32(portNum), Protocol: portProto}
+				subsets, readyEps, notReadyEps = addEndpointSubset(subsets, pod, epa, epp, tolerateUnreadyEndpoints)
```
you want to aggregate `readyEps` and `notReadyEps`, right? 
Otherwise it will always ended up with 1 or 0.

xiangpengzhao replied:
Good catch! will fix it.

freehan reviewed (commented):
Sorry for the late response. Please also squash commits. 
We are almost there. Thanks!

xiangpengzhao commented:
inline comments addressed and commits squashed. PTAL @freehan 
Thanks!

xiangpengzhao commented:
/test pull-kubernetes-unit
flake.

freehan commented on pkg/controller/endpoint/endpoints_controller.go:
```diff
@@ -336,63 +335,62 @@ func (e *EndpointController) syncService(key string) error {
-	readyEps := 0
-	notReadyEps := 0
+	subsets := []v1.EndpointSubset{}
+	var readyEps int = 0
```
nit: rename to totalReadyEps and totalNotReadyEps

freehan commented on pkg/controller/endpoint/endpoints_controller.go:
```diff
@@ -336,63 +335,62 @@ func (e *EndpointController) syncService(key string) error {
+					continue
+				}
+
+				var readyEpsNum, notReadyEpsNum int
```
readyEps and notReadyEps

freehan reviewed (commented):
LGTM overall. just nits.

xiangpengzhao commented:
nit addressed, rebased. PTAL @freehan 
thanks!

freehan commented:
/lgtm
@bowei for /approve

bowei commented:
/approve
